enhancements/machine-config/on-cluster-layering.md

```diff
@@ -0,0 +1,1137 @@
+  ClusterRoleBinding applied without requiring elevated privileges. This
+  was done with an off-the-shelf Buildah image
+  (`quay.io/buildah/stable:latest`) and it was verified to work using the
+  lifecycled approach discussed below.
+
+- One of the promises that OpenShift makes is that all of the software
+  that is included in a given OpenShift release is tested to work
+  together, with no version change until an OpenShift upgrade is
+  performed. This same promise should extend to the image builder. To
+  implement this, Buildah is installed into the MCO image via the ART RPM
+  repository which is lifecycled with OCP. It is possible to install a more
+  recent version of Buildah from a different RPM repository, if desired, to
+  ameliorate concerns around CVEs to Buildah.
+
+- For multiarch cases, we should schedule build pods on each
+  architecture within a given cluster. This reduces the need for
+  elevated privileges and would also make the build process a bit faster
+  since the build is occurring natively as opposed to being emulated via QEMU.
+  
+- Instead of creating bare pods, using a higher-level controller such as
+  [Kubernetes Jobs](https://kubernetes.io/docs/concepts/workloads/
+  controllers/job/) is highly recommended. The rationale is because if the
+  node that the build pod is running on is drained and the build pod
+  evicted, the build process will fail erroneously. Working around this
+  requires a lot of complex retry logic that BuildController should not have
+  to concern itself with. Instead, by leveraging the Kubernetes Jobs
+  controller, all BuildController has to concern itself with is whether the
+  build job succeeded. This also allows us to implement retries to work
+  around transient network issues, etc.
+
+#### Pre-Build Checks
+    
+Prior to beginning the build process, a series of checks should be
+performed. For example, the image pullspec for each image within the
+Containerfile should be extracted and a simple `skopeo inspect` query should
+be performed using the provided base image pull secrets. Doing this will
+allow fast feedback to ascertain whether the image build will be successful.
+Additionally, we could perform some additional checks of the push secret and
+final image pull secrets as well.
+    
+#### Pre-Rollout Checks
+    
+Prior to applying the newly-built image, there are some checks that should
+be considered to validate that the image will work. To be clear, this should
+not be considered a pre-rollout test (as described below). Instead, these
+should be considered a quick sanity check to validate that the image has a
+high probability of working. These checks should include, but are not
+limited to:
+    
+- Check that the Kubelet version is the same.
+- Check that the base OS image is the same.
+- Check that the OS extensions image is the same.
+    
+#### Pre-Rollout Tests
+    
+Within IPI clusters, it might be desirable to have an automated way of
```

Comment:
Why does IPI or UPI make a difference, what features are you relying on for this?